python/ray/util/sgd/v2/trainer.py

```diff
@@ -132,41 +153,149 @@ def run(self,
-            finished_with_errors = True
-            raise RuntimeError(
-                "This Trainer is not active. It is either shutdown already or "
-                "never started in the first place. Either create a new "
-                "Trainer or start this one.") from None
-        except SGDBackendError:
-            finished_with_errors = True
-            raise RuntimeError("Training failed. You should not be seeing "
-                               "this error and this is a bug. Please create "
-                               "a new issue at "
-                               "https://github.com/ray-project/ray.") from None
+        try:
+            iterator = self.run_iterator(
+                train_func=train_func, config=config, checkpoint=checkpoint)
+            for intermediate_result in iterator:
+                for callback in callbacks:
+                    callback.handle_result(intermediate_result)
+
+            assert iterator.is_finished()
+            return iterator.get_returns()
         finally:
             for callback in callbacks:
                 callback.finish_training(error=finished_with_errors)
 
+    def run_iterator(
+            self,
+            train_func: Union[Callable[[], T], Callable[[Dict[str, Any]], T]],
+            config: Optional[Dict[str, Any]] = None,
+            checkpoint: Optional[Dict] = None) -> Iterator[List[Dict]]:
+        """Same as ``run`` except returns an iterator over the results.
+
+        This is useful if you want to have more customization of what to do
+        with the intermediate results or how to use the ``Trainer`` with Ray
+        Tune.
+
+        .. code_block:: python
+
+            def train_func(config):
+                ...
+                for _ in config["epochs"]:
+                    metrics = train()
+                    metrics = validate(...)
+                    ray.sgd.report(**metrics)
+                return model
+
+            iterator = trainer.run_iterator(train_func, config=config)
+
+            for result in iterator:
+                do_stuff(result)
+                latest_ckpt = trainer.get_latest_checkpoint()
+
+            assert iterator.is_finished()
+            model = iterator.get_returns()[0]
+
+        Args:
+            train_func (Callable): The training function to execute.
+                This can either take in no arguments or a ``config`` dict.
+            config (Optional[Dict]): Configurations to pass into
+                ``train_func``. If None then an empty Dict will be created.
+            checkpoint (Optional[Dict]): The checkpoint data that should be
+                loaded onto each worker and accessed by the training function
+                via ``sgd.load_checkpoint()``.
+
+        Returns:
+            An Iterator over the intermediate results from ``sgd.report()``.
+        """
+        train_func = self._get_train_func(train_func, config)
+
+        class SGDIterator:
```

Comment:
BTW, I would consider moving this out (instead of nesting) because it's easier to test

Reply:
Done!